static/app/components/core/checkbox/index.tsx

```diff
@@ -0,0 +1,155 @@
+import {forwardRef, useImperativeHandle, useLayoutEffect, useRef} from 'react';
+import styled from '@emotion/styled';
+
+import InteractionStateLayer from 'sentry/components/interactionStateLayer';
+import type {FormSize} from 'sentry/utils/theme';
+
+type CheckboxConfig = {
+  borderRadius: string;
+  box: string;
+  icon: string;
+};
+
+const checkboxSizeMap: Record<FormSize, CheckboxConfig> = {
+  xs: {box: '12px', borderRadius: '2px', icon: '10px'},
+  sm: {box: '16px', borderRadius: '4px', icon: '12px'},
+  md: {box: '22px', borderRadius: '6px', icon: '18px'},
+};
+
+export interface CheckboxProps
+  extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'checked' | 'size'> {
+  /**
+   * Is the checkbox active? Supports 'indeterminate'
+   */
+  checked?: React.InputHTMLAttributes<HTMLInputElement>['checked'] | 'indeterminate';
+  /**
+   * The size of the checkbox. Defaults to 'sm'.
+   */
+  size?: FormSize;
+}
+
+export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(
+  ({checked = false, size = 'sm', className, ...props}, ref) => {
+    const nativeCheckboxRef = useRef<HTMLInputElement>(null);
+    useImperativeHandle(ref, () => nativeCheckboxRef.current as HTMLInputElement);
+
+    useLayoutEffect(() => {
+      if (nativeCheckboxRef.current) {
+        nativeCheckboxRef.current.indeterminate = checked === 'indeterminate';
+      }
+    }, [checked]);
```

Comment:
if we move this all to a callback ref, I think it would be a bit simpler:

```
const nativeCheckboxRef = useCallback((node: HTMLInputElement | null) => {
  if (node) {
    node.indeterminate = checked === 'indeterminate'
  }
}, [checked])
```

we can then simply pass `nativeCheckboxRef` directly as the `ref` prop, merged with the `ref` from `forwardRef`:

```
ref={mergeRefs([nativeCheckboxRef, ref])}
```

Reply:
Much nicer! I had tried this and then saw it broke tests (I think I wasnt observing the checked value or something)